modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2051,13 +2051,48 @@ def sort_function(df):
         # If this df is empty, we don't want to try and shuffle or sort.
         if len(self.axes[0]) == 0 or len(self.axes[1]) == 0:
             return self.copy()
-        # If this df only has one row partition, we don't want to do a shuffle and sort - we can
-        # just do a full-axis sort.
-        if len(self._partitions) == 1:
-            return self.apply_full_axis(
-                1,
-                sort_function,
-            )
+
+        ideal_num_new_partitions = len(self._partitions)
+        m = len(self.index) / ideal_num_new_partitions
+        sampling_probability = (1 / m) * np.log(
+            ideal_num_new_partitions * len(self.index)
+        )
+        # If this df is overpartitioned, we try to sample each partition with probability
+        # greater than 1, which leads to an error. In this case, we can do one of the following
+        # two things. If there is only enough rows for one partition, and we have only 1 column
+        # partition, we can just combine the overpartitioned df into one partition, and sort that
+        # partition. If there is enough data for more than one partition, we can tell the sorting
+        # algorithm how many partitions we want to end up with, so it samples and finds pivots
+        # according to that.
+        if sampling_probability >= 1:
+            from modin.config import MinPartitionSize
+
+            ideal_num_new_partitions = round(len(self.index) / MinPartitionSize.get())
+            if len(self.index) < MinPartitionSize.get() or ideal_num_new_partitions < 2:
+                modin_frame = self
+                if self._partitions.shape[1] != 1:
```

Comment:
why do we need to manually wrap partitions into a row-parts? Shouldn't `apply_full_axis(axis=1)` do this for us automatically? (self.apply_full_axis -> self.broadcast_apply_full_axis -> [part_mgr.broadcast_axis_partitions](https://github.com/modin-project/modin/blob/db7103b22f1f1efe665a63a42673c491dd148ec8/modin/core/dataframe/pandas/partitioning/partition_manager.py#L466))

Reply:
This is for the case where we have more than one row partition, and more than one column partition. If we just do self.apply_full_axis, we get N partitions (where N was the number of column partitions we had) partitions with all of the rows, so sorting can fail on a KeyError if not every partition has the sort columns. To avoid this, if we have more than one column partition, we combine all of them so every partition has all of the columns and some of the rows, and then do a full axis sort on the other axis (so we get one big partition with all of the data).